src/include/plan_node/aggregate_plan_node.h

```diff
@@ -0,0 +1,203 @@
+      return *this;
+    }
+
+    /**
+     * @param strategy aggregation strategy to be used
+     * @return builder object
+     */
+    Builder &SetAggregateStrategy(AggregateStrategy strategy) {
+      aggregate_strategy_ = strategy;
+      return *this;
+    }
+
+    /**
+     * Build the aggregate plan node
+     * @return plan node
+     */
+    std::shared_ptr<AggregatePlanNode> Build() {
+      return std::shared_ptr<AggregatePlanNode>(
+          new AggregatePlanNode(std::move(children_), std::move(output_schema_), estimated_cardinality_,
+                                std::move(having_clause_predicate_), std::move(aggregate_terms_), aggregate_strategy_));
+    }
+
+   protected:
+    /**
+     * Predicate for having clause if it exists
+     */
+    std::unique_ptr<const parser::AbstractExpression> having_clause_predicate_;
+    /**
+     * List of aggregate terms for aggregation
+     */
+    std::vector<AggregateTerm> aggregate_terms_;
+    /**
+     * Strategy to use for aggregation
+     */
+    AggregateStrategy aggregate_strategy_;
+  };
+
+  /**
+   * @param children child plan nodes
+   * @param output_schema Schema representing the structure of the output of this plan node
+   * @param estimated_cardinality estimated cardinality of output of node
+   * @param having_clause_predicate unique pointer to possible having clause predicate
+   * @param aggregate_terms vector of aggregate terms for the aggregation
+   * @param aggregate_strategy aggregation strategy to be used
+   */
+  AggregatePlanNode(std::vector<std::unique_ptr<AbstractPlanNode>> &&children,
+                    std::shared_ptr<OutputSchema> output_schema, uint32_t estimated_cardinality,
+                    std::unique_ptr<const parser::AbstractExpression> &&having_clause_predicate,
+                    std::vector<AggregateTerm> aggregate_terms, AggregateStrategy aggregate_strategy)
+      : AbstractPlanNode(std::move(children), std::move(output_schema), estimated_cardinality),
+        having_clause_predicate_(std::move(having_clause_predicate)),
+        aggregate_terms_(std::move(aggregate_terms)),
+        aggregate_strategy_(aggregate_strategy) {}
+
+ public:
+  ~AggregatePlanNode() override {
+    for (auto term : aggregate_terms_) {
+      delete term.expression_;
+    }
+  }
+
+  //===--------------------------------------------------------------------===//
+  // ACCESSORS
+  //===--------------------------------------------------------------------===//
+
+  // TODO(Gus,Wen): Figure out how to represent global aggregates (example: count(*))
+  /**
+   * A global aggregate does not aggregate over specific columns, example: count(*)
+   * @return true if aggregation is global
+   */
+  bool IsGlobal() const { return false; }
+
+  /**
+   * @return pointer to predicate for having clause
+   */
+  const parser::AbstractExpression *GetHavingClausePredicate() const { return having_clause_predicate_.get(); }
```

Comment:
What's the point of having a unique pointer if we're just handing out the raw pointer?

Reply:
Unsure, seems like a disease from the old code. Changed it to return the unique pointer